tests/xfstests/run.pm

```diff
@@ -0,0 +1,176 @@
+}
+
+# Stop heartbeat
+sub heartbeat_stop {
+    type_string("\n");
+    assert_script_run("touch $HB_EXIT_FILE");
+}
+
+# Wait for heartbeat
+sub heartbeat_wait {
+    my $ret = wait_serial([$HB_PATN, $HB_DONE], $HB_TIMEOUT);
+    if ($ret) {
+        if ($ret =~ /$HB_PATN/) {
+            return ($HB_PATN, "");
+        }
+        else {
+            my $status;
+            type_string("\n");
+            my $ret = script_output("cat $HB_DONE_FILE; rm -f $HB_DONE_FILE");
+            $ret =~ s/^\s+|\s+$//g;
+            if ($ret == 0) {
+                $status = "PASSED";
+            }
+            elsif ($ret == 22) {
+                $status = "SKIPPED";
+            }
+            else {
+                $status = "FAILED";
+            }
+            return ($HB_DONE, $status);
+        }
+    }
+    return ("", "FAILED");
+}
+
+# Add one test result to log file
+sub log_add {
+    my $file   = shift;
+    my $name   = shift;
+    my $status = shift;
+    my $time   = shift;
+    my $cmd    = "echo '$name ... ... $status (${time}s)' >> $file";
+    type_string("\n");
+    assert_script_run($cmd);
+}
+
+# Create log dir and set environment variables
+sub test_prepare {
+    my $category = shift;
+    type_string("mkdir -p /tmp/$category; source ~/.xfstests\n");
+}
+
+# List all the tests of a specific category
+sub test_list {
+    my $dir    = shift;
+    my $output = script_output("find $dir -regex '.*/[0-9]+'", 200);
+    my @tests  = split(/\n/, $output);
+    foreach my $test (@tests) {
+        $test = basename($test);
+    }
+    return @tests;
+}
+
+# Run a single test and write log to file
+# category - category of tests(e.g. xfs, btrfs, generic)
+# test  - specific test(e.g. 001, 002)
+sub test_run {
+    my $category = shift;
+    my $test     = shift;
+    my $cmd      = "\n$WRAPPER '$category/$test' | ";
+    $cmd .= "tee $HB_RAMFS/test.log; ";
+    $cmd .= "echo \${PIPESTATUS[0]} > $HB_DONE_FILE; ";
+    $cmd .= "mv $HB_RAMFS/test.log /tmp/$category/$test.log\n";
+    type_string($cmd);
+}
+
+
+sub run {
+    my $self = shift;
+    select_console('root-console');
+
+    # Get test list
+    my ($filesystem, $category) = split(/-/, get_var("XFSTESTS"));
+    my @tests = test_list("/opt/xfstests/tests/$category");
+
+    heartbeat_start;
+    test_prepare($category);
+    my @crashed;
+    foreach my $test (@tests) {
+        if (exists($BLACKLIST{"$category/$test"})) {
+            next;
+        }
+
+        my $name = "$category-$test";
+        my $time = $HB_INTVL;
+        # Run test
+        test_run($category, $test);
+
+        # Waiting for test to finish
+        my ($type, $status) = heartbeat_wait;
+        while ($type eq $HB_PATN) {
+            ($type, $status) = heartbeat_wait;
+            $time += $HB_INTVL;
+        }
+        if (!$type) {
+            # System crash
+            power_action('reboot', observe => 1, keepconsole => 1);
+            $self->wait_boot;
+            sleep(5);
```

Comment:
looks strange between wait_boot and select_console , both function are capable to wait without additional timeout

Reply:
During testing on our local openQA, after wait_boot, sometimes select_console failed to get the terminal. I don't have time to debug it. "sleep" is an easy workaround.